metadata-ingestion/src/datahub/cli/generate_report/report_generator.py

```diff
@@ -0,0 +1,247 @@
+    GRAPHQL_QUERY: str = """
+    query GetDatasetGlossaryTerms($search_query: String!, $page_size: Int, $offset:Int) {
+      search(input: {
+        type: DATASET,
+        # search query cannot be empty, use "*" or by searching for a specific dataPlatoform name e.g. snowflake
+        query: $search_query,
+        count: $page_size,
+        start: $offset,
+      }) {
+        total
+        searchResults {
+          entity {
+            ... on Dataset {
+              urn
+              editableSchemaMetadata {
+                editableSchemaFieldInfo {
+                  fieldPath
+                  glossaryTerms {
+                    terms {
+                      term {
+                        urn
+                        name
+                      }
+                    }
+                  }
+                }
+              }
+              # zero represents the latest version
+              schemaMetadata(version: 0) {
+                name
+                fields {
+                  fieldPath
+                  glossaryTerms {
+                    terms {
+                      term {
+                        urn
+                        name
+                      }
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+    """
+    graphql_api_url: str
+
+    def __init__(self, datahub_base_url: str) -> None:
+        self.graphql_api_url = f'{datahub_base_url}{self.DATAHUB_GRAPHQL_ENDPOINT}'
+
+    def yield_search_results(self, search_queries: Iterable[str]) -> Generator[Dict, None, None]:
+        return itertools.chain.from_iterable(
+            self._yield_rows(entity)
+            for search_query in search_queries
+            for entity in self._yield_graphql_search_results(search_query)
+        )
+
+    def _yield_graphql_search_results(self, search_query: str) -> Generator[Dict, None, None]:
+        total = self.MAX_SEARCH_RESULTS_TOTAL
+        offset = 0
+        while offset < total:
+            graphql_variables = {
+                'search_query': search_query,
+                'page_size': self.PAGE_SIZE,
+                'offset': offset
+            }
+            logger.info(f'Issuing graphql query with variables: {graphql_variables}')
+
+            payload = {'query': self.GRAPHQL_QUERY, 'variables': graphql_variables}
+            response = requests.post(self.graphql_api_url, json=payload)
+            response.raise_for_status()
+            response_json = response.json()
+            for entity in response_json['data']['search']['searchResults']:
+                yield entity
+
+            total = response_json['data']['search']['total']
+            if total > self.MAX_SEARCH_RESULTS_TOTAL:
+                msg = f'graphql query total results ({total}) exceeds maximum ({self.MAX_SEARCH_RESULTS_TOTAL})'
+                raise RuntimeError(msg)
+            offset += self.PAGE_SIZE
```

Comment:
Is it safe to always increment the offset by `PAGE_SIZE`? Ideally we are guaranteed to get back `PAGE_SIZE` items if there are >= `PAGE_SIZE` items remaining, but if they're treating this as a "max" page size (or do so in the future), they could be within their rights to return less than the size we specified.

Ex:
- `PAGE_SIZE` = 500
- Execute API for offset 0
- Datahub returns 300 items, but still has 700 other items to process
- We (incorrectly) execute API for offset 500, when we should've used 300, causing us to miss 200 items

Just to be on the safe side, would it be better to increment by the actual number of items returned, rather than our expectation?

Reply:
right, i assume that it will return `PAGE_SIZE = 500` if there are actually more than that.

> Just to be on the safe side, would it be better to increment by the actual number of items returned, rather than our expectation?

Yeah I can change that.

Reply:
done https://github.com/Affirm/datahub/pull/3/commits/3deac44ed9124d35c1364a60d1e1a94ea75de193#diff-756d114d7c20e83be0185b88aa5166cfc412702b0e1404a6a46e06ff798904b0R133-R139